fvm/evm/stdlib/contract.go

```diff
@@ -38,6 +46,790 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+		}
+
+	case *interpreter.ArrayValue:
+		arrayStaticType := value.Type
+
+		arrayGethABIType, ok := gethABIType(arrayStaticType, evmAddressTypeID)
+		if !ok {
+			break
+		}
+
+		elementStaticType := arrayStaticType.ElementType()
+
+		elementGoType, ok := goType(elementStaticType, evmAddressTypeID)
+		if !ok {
+			break
+		}
+
+		var result reflect.Value
+
+		switch arrayStaticType := arrayStaticType.(type) {
+		case interpreter.ConstantSizedStaticType:
+			size := int(arrayStaticType.Size)
+			result = reflect.Indirect(reflect.New(reflect.ArrayOf(size, elementGoType)))
+
+		case interpreter.VariableSizedStaticType:
+			size := value.Count()
+			result = reflect.MakeSlice(reflect.SliceOf(elementGoType), size, size)
+		}
+
+		var index int
+		value.Iterate(inter, func(element interpreter.Value) (resume bool) {
+
+			arrayElement, _, err := encodeABI(
+				inter,
+				locationRange,
+				element,
+				element.StaticType(inter),
+				evmAddressTypeID,
+			)
+			if err != nil {
+				panic(err)
+			}
+
+			result.Index(index).Set(reflect.ValueOf(arrayElement))
+
+			index++
+
+			// continue iteration
+			return true
+		})
+
+		return result.Interface(), arrayGethABIType, nil
+	}
+
+	return nil, gethABI.Type{}, abiEncodingError{
+		Type: value.StaticType(inter),
+	}
+}
+
+// EVM.decodeABI
+
+const internalEVMTypeDecodeABIFunctionName = "decodeABI"
+
+var internalEVMTypeDecodeABIFunctionType = &sema.FunctionType{
+	Parameters: []sema.Parameter{
+		{
+			Identifier: "types",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.NewVariableSizedType(nil, sema.MetaType),
+			),
+		},
+		{
+			Label:          "data",
+			TypeAnnotation: sema.NewTypeAnnotation(sema.ByteArrayType),
+		},
+	},
+	ReturnTypeAnnotation: sema.NewTypeAnnotation(
+		sema.NewVariableSizedType(nil, sema.AnyStructType),
+	),
+}
+
+func newInternalEVMTypeDecodeABIFunction(
+	gauge common.MemoryGauge,
+	location common.AddressLocation,
+) *interpreter.HostFunctionValue {
+	evmAddressTypeID := location.TypeID(gauge, evmAddressTypeQualifiedIdentifier)
+
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		internalEVMTypeDecodeABIFunctionType,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			inter := invocation.Interpreter
+			locationRange := invocation.LocationRange
+
+			// Get `types` argument
+
+			typesArray, ok := invocation.Arguments[0].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			// Get `data` argument
+
+			dataValue, ok := invocation.Arguments[1].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			invocation.Interpreter.ReportComputation(
+				environment.ComputationKindEVMDecodeABI,
+				uint(dataValue.Count()),
+			)
+
+			data, err := interpreter.ByteArrayValueToByteSlice(inter, dataValue, locationRange)
+			if err != nil {
+				panic(err)
+			}
+
+			var arguments gethABI.Arguments
+			typesArray.Iterate(inter, func(element interpreter.Value) (resume bool) {
```

Comment:
do we need any protection against depth attacks here?

Reply:
See https://github.com/onflow/flow-go/pull/5146